### PR TITLE
safer indexing for AbstractArray

### PR DESCRIPTION
**Pull request by palday**
~closes # 431~

apparently not on Mac

### CODE REVIEW

File: src/convert/base.jl [hunk 1]
```diff
@@ -264,8 +264,8 @@ sexp(::Type{RClass{:character}},st::AbstractString) = sexp(RClass{:character}, s
 function sexp(::Type{RClass{:character}}, a::AbstractArray{T}) where T<:AbstractString
     ra = protect(allocArray(StrSxp, size(a)...))
     try
-        for i in 1:length(a)
-            ra[i] = a[i]
+        for (i, idx) in enumerate(eachindex(a))
+            ra[i] = a[idx]
```
Comment (ararslan) [inline src/convert/base.jl:268]: Note that this still assumes 1-based indexing for `ra` since `enumerate` will start with 1.

Comment (palday) [reply src/convert/base.jl:268]: yes, but we create `ra` and it's indeed one based

Comment (ararslan) [reply src/convert/base.jl:268]: Fair, though we also create `a` and `v` as 1-based arrays in the above `rcopy` methods which means the loop specifications there were safe prior to using `eachindex`. I assumed you were going for minimal assumptions but if not then that's fine.

**Comment by ararslan** [comment]
```
MethodError: no method matching keys(::Ptr{StrSxp})
```
Looks like something is getting `eachindex`ed that shouldn't be (or is missing a suitable definition, perhaps)

File: src/convert/base.jl (hunk 1 above)
Comment (palday) [reply src/convert/base.jl:270]: `typeof(ra)` is lacking a `eachindex` method and the fallback calls `keys`; I'm going to take a stab at defining something appopriate

File: src/methods.jl [hunk 2]
```diff
@@ -99,6 +99,7 @@ function iterate(s::Ptr{S}, state) where S<:VectorSxp
     (s[state], state)
 end
 
+Base.eachindex(s::Ptr{<:VectorSxp}) = Base.OneTo(length(s))
```
Comment (ararslan) [inline src/methods.jl:102]: Feels weird to define `eachindex` for a `Ptr` but we also have `iterate` and `getindex`...

File: src/convert/base.jl [hunk 3]
```diff
@@ -296,8 +298,10 @@ end
 function sexp(::Type{RClass{:list}}, a::AbstractArray)
     ra = protect(allocArray(VecSxp, size(a)...))
     try
-        for i in 1:length(a)
-            ra[i] = a[i]
+        # we want this to work even if a doesn't use one-based indexing
+        # we only care about ra having the same length (which it does)
+        for (i, idx) in zip(eachindex(ra), eachindex(a))
+            ra[i] = a[idx]
```
Comment (ararslan) [inline src/convert/base.jl:304]: This feels like a situation where a `copyto!` method could make sense, assuming such a method doesn't already exist (I assume it doesn't)

**Review by ararslan — APPROVED** [review]

File: src/methods.jl (hunk 2 above)
Comment (palday) [reply src/methods.jl:102]: yeah, it does, but then again we're treating R native arrays as iterators without copying...

File: src/convert/base.jl (hunk 3 above)
Comment (palday) [reply src/convert/base.jl:304]: Hmmm, yeah, but I'm going to punt on that until I've tackled #444

**Pull request merged**